Simple payments: allow read+write access to meta fields via REST API

simison opened the pull request:
Enable Simple payments custom meta values for access through the REST API. Field’s value will be exposed on a `.meta` key in the endpoint response, and WordPress will handle setting up the callbacks for reading and writing to that meta key.

This is needed for Simple Payments block to be able to read+create+update Simple payment buttons through the REST API.

https://developer.wordpress.org/rest-api/extending-the-rest-api/modifying-responses/
https://developer.wordpress.org/reference/functions/register_meta/

## Changes proposed in this Pull Request:

Enable Simple payments custom meta values for access through the REST API.

## Testing instructions:

- Have a site with simple payments enabled (premium or business plan)
- Create a post in Calypso editor and create a simple payments button (from "+ add" menu)
- Note down button's ID by looking at the posts HTML source (from shortcode e.g. `[simple-payment id="42"]`)
- Call API endpoint: 
    - Locally hosted Jetpack
        ```
        http://localhost/index.php?rest_route=/wp/v2/jp_pay_product/BUTTON_ID
        ```
    - WPCOM (apply D19907-code to test):
        ```
        https://public-api.wordpress.com/wp/v2/sites/YOUR_SITE_URL/jp_pay_product/BUTTON_ID
        ```
- Without this change, `meta` field is missing Simple payments meta fields, with the change you should see:
```
  "meta": {
    "spay_price": 3,
    "spay_currency": "HUF",
    "spay_cta": "",
    "spay_multiple": 1,
    "spay_email": "email@example.com",
    "spay_formatted_price": "Ft3",
    "spay_status": ""
  },
```

You should be able to write to this object as well. The easiest way to test reading+creating+writing would be following instructions from https://github.com/Automattic/wp-calypso/pull/28065

## Proposed changelog entry for your changes:

Enable Simple payments custom meta values for access through the REST API.

simison committed:
Simple payments: allow read+write access to meta fields via REST API

Enable Simple payments custom meta values for access through the REST API. Field’s value will be exposed on a .meta key in the endpoint response, and WordPress will handle setting up the callbacks for reading and writing to that meta key.

https://developer.wordpress.org/rest-api/extending-the-rest-api/modifying-responses/
https://developer.wordpress.org/reference/functions/register_meta/

kraftbj commented on modules/simple-payments/simple-payments.php:
```diff
@@ -40,6 +40,9 @@ private function register_scripts_and_styles() {
 
 	private function register_init_hook() {
 		add_action( 'init', array( $this, 'init_hook_action' ) );
+		if ( function_exists( 'register_meta' ) ) {
```
What's the purpose of checking for this function? When would you expect it not to be present?

simison replied:
Just played safe without checking how long it has been available. Seems like since WordPress 3.3.0 so it's safe to remove the check.

https://developer.wordpress.org/reference/functions/register_meta/

rodrigoi commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +289,67 @@ function allow_sync_post_meta( $post_meta ) {
+	 * @link https://developer.wordpress.org/reference/functions/register_meta/
+	 */
+	public function register_meta_fields_in_rest_api() {
+			register_meta( 'post', 'spay_price', array(
```
not required, but we can add:

```php
'object_subtype' => self::$post_type_product,
```

so we don't add these to all posts.

simison replied:
Added! Good catch.

So in WordPress 4.9.8, `object_subtype` [was added](https://make.wordpress.org/core/2018/07/27/registering-metadata-in-4-9-8/) together with new shorthand helper:

```php
register_post_meta( 'jp_pay_product', 'spay_price', array(
    ...
) );
```

Is the same as:
```php
register_meta( 'post', 'spay_price', array(
    'object_subtype' => 'jp_pay_product',
    ...
) );
```

...with the difference that prior to WordPress 4.9.8 where `object_subtype` isn't supported yet, the latter will expose`spay_price` meta field in REST API endpoints for ALL post types, not just for `jp_pay_product` custom post type.

Should we therefore rather `register_post_meta` and not have functioning API for Simple Payments block prior to WP 4.9.8 at all, or expose these fields via all post types in older WP versions. Suppose our meta fields have unique enough names (and not "price" or "email") so that we wouldn't accidentally expose wrong meta fields from wrong posts.

simison committed:
Update register_meta options for each field

simison committed:
Don't check if `register_meta` function exists

...since it's been there since WordPress 3.3.0

https://developer.wordpress.org/reference/functions/register_meta/

jeherve reviewed (dismissed):
Seems to work well for me. 🚢

tyxla commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +287,78 @@ function allow_sync_post_meta( $post_meta ) {
+			register_meta( 'post', 'spay_price', array(
+				'description'       => esc_html__( 'Simple payments; price.', 'jetpack' ),
+				'object_subtype'    => self::$post_type_product,
+				'sanitize_callback' => 'absint',
```
Noting that `absint()` will throw away any non-integer value from the price - is that intentional? I'd expect that we support two digits after decimal`0.00` for prices. We're probably doing this intentionally (as I see it's enforced in the UI as well), but I'd love to know why.

simison replied:
Good point — there's some confusion about price since the current button in Calypso complains about using digits but placeholder text suggests they're ok.

@rodrigoi can you elaborate on spec of these values a bit, especially the price has been difficult to navigate.

simison replied:
This now has a proper price and currency validation.

I'm not validating differently based on different currencies since that seems quite tricky to do here. (Three currencies don't allow decimals). We are validating for that in the client and PayPal will throw an error when attempting to use it with such price.

rodrigoi replied:
I'll mark this as _resolved_ because, as @simison pointed out, as long as the block UI has some validation, the payment processor will catch those and report them as errors.

tyxla reviewed (dismissed):
I'm testing with #28065 and this seems to work great! Nice work!

Left some minor comments, I'm mostly concerned about our lack of support for non-integer values for prices.

simison committed:
Fix intendation

jeherve reviewed (dismissed):
Adding my 🚢 👍 back. Merge when ready.

simison commented:
I'm a bit cautions to merge without fully understanding implications of just casting `spay_price` as an Integer or a Float (convo p1540815622091600-slack-jetpack-gutenberg) — I'mg going to have a better look with more time tomorrow.

tyxla commented:
Added to 6.8 milestone, we want it in :)

simison committed:
Add price and currency validation

rodrigoi commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +287,134 @@ function allow_sync_post_meta( $post_meta ) {
+			'type'              => 'string',
+		) );
+
+		register_meta( 'post', 'spay_formatted_price', array(
```
As I've mentioned on https://github.com/Automattic/wp-calypso/pull/28169#discussion_r230430737 we are safe to remove this `meta`. I'm working on removing it permanently so there's no more confusion with `spay_price`.

sirreal replied:
👍 Let's drop this, I wasn't involved in the conversation but having the client push a formatted price raised some red flags for me when I was initially looking at this.

simison replied:
Removed in https://github.com/Automattic/jetpack/pull/10402/commits/3e393a3c57f0c810ac218a4cb322dcc123265f2e

rodrigoi replied:
❤️

sirreal commented on modules/simple-payments/simple-payments.php:
```diff
@@ -40,6 +40,7 @@ private function register_scripts_and_styles() {
 
 	private function register_init_hook() {
 		add_action( 'init', array( $this, 'init_hook_action' ) );
+		add_action( 'rest_api_init', array( $this, 'register_meta_fields_in_rest_api' ) );
```
NIT: Function name is now a bit misleading, it registers `init` and another hook.

simison replied:
Do you mean `register_init_hook` is misleading or that `register_meta_fields_in_rest_api` is misleading?

I renamed it to plural `register_init_hook` → `register_init_hooks` form just now)

Both actions are called `init`s, the other one just for the rest-API.

sirreal replied:
This is fine, no change required really.

This was more of an observation that method names like this which could be more generic become outdated as soon as we need more and different functionality.

sirreal commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +287,134 @@ function allow_sync_post_meta( $post_meta ) {
+	 * List has to be in sync with list at the client side:
+	 * @link https://github.com/Automattic/wp-calypso/blob/6d02ffe73cc073dea7270a22dc30881bff17d8fb/client/lib/simple-payments/constants.js
+	 */
+	public function sanitize_currency( $currency ) {
```
Any reason this isn't static? `sanitize_price` is. Let's be consistent.

sirreal commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +287,134 @@ function allow_sync_post_meta( $post_meta ) {
+		register_meta( 'post', 'spay_multiple', array(
+			'description'       => esc_html__( 'Simple payments; allow multiple items', 'jetpack' ),
+			'object_subtype'    => self::$post_type_product,
+			'sanitize_callback' => 'absint',
```
I understand that this is a boolean, should we use [`rest_sanitize_boolean`](https://developer.wordpress.org/reference/functions/rest_sanitize_boolean/)?

sirreal commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +287,134 @@ function allow_sync_post_meta( $post_meta ) {
+	 * @link https://developer.paypal.com/docs/integration/direct/rest/currency-codes/
+	 *
+	 * List has to be in sync with list at the client side:
+	 * @link https://github.com/Automattic/wp-calypso/blob/6d02ffe73cc073dea7270a22dc30881bff17d8fb/client/lib/simple-payments/constants.js
```
Probably worth noting that both should align with PayPal (has a comment there):

https://developer.paypal.com/docs/integration/direct/rest/currency-codes/


https://github.com/Automattic/wp-calypso/blob/6d02ffe73cc073dea7270a22dc30881bff17d8fb/client/lib/simple-payments/constants.js#L4

sirreal reviewed (commented):
Nice work here. In particular, the documentation with links is excellent 🙌 

Let's be sure to add a comment in Calypso as well that currencies should be kept in sync:

https://github.com/Automattic/wp-calypso/blob/6d02ffe73cc073dea7270a22dc30881bff17d8fb/client/lib/simple-payments/constants.js#L4

I've tested and this seems to work well 👍 

I've left a few minor notes, but this could probably ship any time.

simison committed:
Remove `spay_formatted_price`

simison committed:
static sanitize_currency

simison committed:
Treat multiple as boolean

simison committed:
register_init_hook → register_init_hooks

simison committed:
Clarify `sanitize_currency` doc

sirreal commented on modules/simple-payments/simple-payments.php:
```diff
@@ -286,6 +287,126 @@ function allow_sync_post_meta( $post_meta ) {
+	 * @link https://developer.paypal.com/docs/integration/direct/rest/currency-codes/
+	 */
+	public static function sanitize_currency( $currency ) {
+		$valid_currencies = array(
```
See https://github.com/Automattic/wp-calypso/pull/28236

sirreal commented:
Rebased.

I added a change that HTML formats the price for display on the fly which has been extracted to #10555.

The change addresses the problem mentioned in D19907-code. Since we don't send the formatted price, now we need to render a formatted HTML price to avoid a discrepancy between the actual price and the displayed price:

![screen shot 2018-11-06 at 13 07 39](https://user-images.githubusercontent.com/841763/48130958-9baf7200-e28e-11e8-9dca-181b32b230ec.png)

jeherve reviewed (dismissed):
This seems to work well for me. 👍

sirreal commented:
I've rebased this without the changes in #10555, which have changed some and are significant enough to warrant an independent PR.

sirreal reviewed (approved):
This was working well. I'm re-approving after removing some formatting commits that are now part of #10555

jeherve reviewed (approved):
Works for me. Merge when ready.

jeherve commented:
Cherry-picked to `branch-6.7` in 1c7176e3320ac4ec17aa34222b064e06dab29742
